File: backend/app/security/permissions.py
from enum import StrEnum


class Permission(StrEnum):
    SOC_READ = "soc:read"
    INVESTIGATIONS_CREATE = "investigations:create"
    USERS_MANAGE = "users:manage"
    API_KEYS_MANAGE = "api_keys:manage"
    AUDIT_READ = "audit:read"


ROLE_PERMISSIONS: dict[str, frozenset[Permission]] = {
    "viewer": frozenset({Permission.SOC_READ}),
    "analyst": frozenset({Permission.SOC_READ, Permission.INVESTIGATIONS_CREATE}),
    "admin": frozenset(Permission),
}


def permissions_for_role(role: str) -> frozenset[Permission]:
    return ROLE_PERMISSIONS.get(role, frozenset())

File: frontend/src/lib/auth/permissions.ts
import type { RoleName } from "@/lib/api/types";

export type Permission =
  | "soc:read"
  | "investigations:create"
  | "users:manage"
  | "api_keys:manage"
  | "audit:read";

const ROLE_PERMISSIONS: Record<RoleName, ReadonlySet<Permission>> = {
  viewer: new Set(["soc:read"]),
  analyst: new Set(["soc:read", "investigations:create"]),
  admin: new Set([
    "soc:read",
    "investigations:create",
    "users:manage",
    "api_keys:manage",
    "audit:read",
  ]),
};

export function hasPermission(role: RoleName, permission: Permission) {
  return ROLE_PERMISSIONS[role].has(permission);
}

export function canAccessAdmin(role: RoleName) {
  return hasPermission(role, "users:manage");
}
